### PR TITLE
Make PAA trust store configurable

### LINKED ISSUE #11913: Refactor AttestationVerifier to employ an AttestationTrustStore interface
#### Problem

Current example implementation of DeviceAttestationVerifier employs a very simplistic built-in PAA root store. This root store should be abstracted to let attestation verifiers get a list in a platform-specific way.

 #### Proposed Solution

- Create a AttestationTrustStore interface used by the DeviceAttestationVerifier interface. Require capabilities:
  - Look-up PAA by SKID
  - Determine if a PAA is revoked, by SKID
  - Determine if a PAI is revoked, by TBD
  - Determine if a DAC is revoked, by TBD
  - Interface should be asynchronous (since it's IO that may include look-ups)
- Implement a "fixed trust store" version (i.e. the current example code with known single dev root)
- Implement a POSIX version that can search through DER files in a particular directory for PAAs

### PR DESCRIPTION
**Pull request by tcarmelveilleux**
#### Problem
PAA store used by DefaultDeviceAttestationVerifier could not be
replaced, forcing a few fixed test roots to always be used and
nothing else, unless completely forking the
DefaultDeviceAttestationVerifier.

Fixes #11913

#### Change overview

- This PR introduces the `AttestationTrustStore` interface, which the
  default `DeviceAttestationVerifier` expects to get configured
  at construction.
- Examples were modified to use the default test PAA root store
- Unit tests updated to use the testing root store
- Refactored simple array-based Root store to self-extract
  the SKID

#### Testing
Testing done: added new units tests which pass, ran cert tests,
validated attestation succeeds the same as before with test keys.

### CODE REVIEW

File: src/credentials/DeviceAttestationVerifier.h [hunk 1]
```diff
@@ -99,6 +99,81 @@ struct DeviceInfoForAttestation
...
+
+        for (paaIdx = 0; paaIdx < mNumCerts; ++paaIdx)
+        {
+            uint8_t skidBuf[Crypto::kSubjectKeyIdentifierLength] = { 0 };
+            candidate                                            = mDerCerts[paaIdx];
+            MutableByteSpan candidateSkidSpan{ skidBuf };
+            VerifyOrReturnError(CHIP_NO_ERROR == Crypto::ExtractSKIDFromX509Cert(candidate, candidateSkidSpan),
+                                CHIP_ERROR_INTERNAL);
+
+            printf("data: %p, size: %zu\n", skid.data(), skid.size());
```
Comment (msandstedt) [inline src/credentials/DeviceAttestationVerifier.h:161]: Should this be calling a chip logging macro?

Comment (andy31415) [inline src/credentials/DeviceAttestationVerifier.h:161]: is this debug code? Should we use ChipLog messages instead or delete?

File: src/credentials/DeviceAttestationVerifier.h [hunk 2]
```diff
@@ -99,6 +99,81 @@ struct DeviceInfoForAttestation
...
+                return CopySpanToMutableSpan(candidate, outPaaDerBuffer);
+            }
+        }
+
+        return CHIP_ERROR_CA_CERT_NOT_FOUND;
+    }
+
+protected:
+    const ByteSpan * mDerCerts;
+    size_t mNumCerts;
```
Comment (andy31415) [inline src/credentials/DeviceAttestationVerifier.h:174]: nit: mNumCerts could also be const.

**Review by andy31415 — APPROVED** [review]

**Comment by andy31415** [comment]
fast track pending review comments: change is on the smaller side and has significant test coverage, written by a domain owner.

File: src/credentials/DeviceAttestationVerifier.h (hunk 1 above)
Comment (tcarmelveilleux) [reply src/credentials/DeviceAttestationVerifier.h:161]: Oops, that's remaining debugging I missed to remove, will remove.

File: src/credentials/DeviceAttestationVerifier.h [hunk 3]
```diff
@@ -99,6 +99,81 @@ struct DeviceInfoForAttestation
...
+     *          span, CHIP_ERROR_CA_CERT_NOT_FOUND if no PAA found that matches `skid.
+     *
+     */
+    virtual CHIP_ERROR GetProductAttestationAuthorityCert(const ByteSpan & skid, MutableByteSpan & outPaaDerBuffer) const = 0;
+};
+
+class ArrayPaaRootStore : public PaaRootStore
+{
+public:
+    explicit ArrayPaaRootStore(const ByteSpan * derCerts, size_t numCerts) : mDerCerts(derCerts), mNumCerts(numCerts) {}
```
[the remark this answers is not in the document]
Comment (tcarmelveilleux) [reply src/credentials/DeviceAttestationVerifier.h]: Clarified

File: src/credentials/DeviceAttestationVerifier.h (hunk 1 above)
Comment (tcarmelveilleux) [reply src/credentials/DeviceAttestationVerifier.h]: Removed

File: examples/chip-tool/commands/common/CHIPCommand.cpp [hunk 4]
```diff
@@ -45,7 +45,10 @@ CHIP_ERROR CHIPCommand::Run()
     chip::Platform::ScopedMemoryBuffer<uint8_t> rcac;
 
     chip::Credentials::SetDeviceAttestationCredentialsProvider(chip::Credentials::Examples::GetExampleDACProvider());
-    chip::Credentials::SetDeviceAttestationVerifier(chip::Credentials::GetDefaultDACVerifier());
+
+    // TODO: Replace testingRootStore with a PaaRootStore that has the necessary official PAA roots available
+    const chip::Credentials::PaaRootStore * testingRootStore = chip::Credentials::GetTestPaaRootStore();
```
[the remark this answers is not in the document]
Comment (tcarmelveilleux) [reply examples/chip-tool/commands/common/CHIPCommand.cpp]: Renamed to AttestationTrustStore

File: src/credentials/DeviceAttestationVerifier.h (hunk 2 above)
Comment (tcarmelveilleux) [reply src/credentials/DeviceAttestationVerifier.h]: Done.

**Review by msandstedt — APPROVED** [review]

**Review by vijs — APPROVED** [review]

**Comment by tcarmelveilleux** [comment]
Removed fast track label since sufficient actual reviewers went through the PR

**Pull request merged**